koap-core/src/commonMain/kotlin/Message.kt

```diff
@@ -580,20 +580,86 @@ sealed class Message {
         data class Oscore(
-            val value: ByteArray,
+            val partialIv: ByteArray,
+            val kidContext: ByteArray?,
+            val kid: ByteArray?,
         ) : Option() {
             init {
+                val value = optionValue()
                 require(value.size in OSCORE_LENGTH_RANGE) {
                     "Oscore length of ${value.size} is outside allowable range of $OSCORE_LENGTH_RANGE"
                 }
             }
 
             override fun equals(other: Any?): Boolean =
-                this === other || (other is Oscore && value.contentEquals(other.value))
+                this === other ||
+                    (
+                        other is Oscore &&
+                            partialIv.contentEquals(other.partialIv) &&
+                            kidContext.contentEquals(other.kidContext) &&
+                            kid.contentEquals(other.kid)
+                    )
 
-            override fun hashCode(): Int = value.contentHashCode()
+            override fun hashCode(): Int {
+                var result = partialIv.contentHashCode()
+                result = 31 * result + kidContext.contentHashCode()
+                result = 31 * result + kid.contentHashCode()
+                return result
+            }
+
+            override fun toString(): String {
+                val args = listOfNotNull(
+                    if (partialIv.size > 0) "partialIv=${partialIv.toHexString()}" else null,
+                    if (kidContext != null) "kidContext=${kidContext.toHexString()}" else null,
+                    if (kid != null) "kid=${kid?.toHexString()}" else null,
+                ).joinToString(separator = ", ")
+                return "Oscore($args)"
+            }
 
-            override fun toString(): String = "Oscore(value=${value.toHexString()})"
+            fun optionValue(): ByteArray {
+                val partialIvLength = partialIv.size
+                if (partialIvLength > 5) {
+                    throw IllegalArgumentException("Partial IV length of $partialIvLength is outside allowable range of 0..5")
+                }
+                val kidContextLength = if (kidContext != null) kidContext.size else 0
+                val kidFlag = if (kid != null) 0x08 else 0
+                val kidContextFlag = if (kidContext != null) 0x10 else 0
+                val flagBits = partialIvLength or kidFlag or kidContextFlag
+                if (flagBits == 0) {
+                    // If the OSCORE flag bits are all zero (0x00), the option value SHALL be empty
+                    return byteArrayOf()
+                }
+                var value = byteArrayOf(flagBits.toByte()) + partialIv
+                if (kidContext != null) {
+                    value += byteArrayOf(kidContextLength.toByte()) + kidContext
+                }
+                if (kid != null) {
+                    value += kid
+                }
+                return value
```

Comment:
these are some 20+17 lines of fairly dense/complicated code, with magic flag constants, etc.
maybe they should go in their own `.kt` file to isolate the complexity, and using constants for flags and masks?
not too happy that `kidFlag` and `kidContextFlag` is a byte/int value in one place and boolean in the other, maybe they should at least be named differently… (`kidMask`/`kidContextMask`?)

Reply:
also, as mentioned in #333 

> There's reserved values for partialIv length 6 and 7, and three flags reserved for future use, so there's some potential for future updates to the standard rendering the option unparseable, and such flags/values should be handled in specific ways as outlined in [8.2](https://www.rfc-editor.org/rfc/rfc8613.html#section-8.2). Verifying the Request …
>
> It could have a malformed flag in it, or maybe a nullable partialIv and use null there to indicate the option didn't parse? I guess these complex options are a bit trickier than the simple uint/string ones in basic CoAP.

not sure how likely the option is to change, and it could just be handled by updating the code. but som drawback compared to just a binary blob is that it might start failing if updated.

Reply:
> these are some 20+17 lines of fairly dense/complicated code, with magic flag constants, etc. maybe they should go in their own `.kt` file to isolate the complexity, and using constants for flags and masks? not too happy that `kidFlag` and `kidContextFlag` is a byte/int value in one place and boolean in the other, maybe they should at least be named differently… (`kidMask`/`kidContextMask`?)

added `Oscore.kt` with constants for flags, masks, and ranges, and updated to `kid[Context]FlagBit`